addon/mixins/scoped-query-cache.js

```diff
@@ -68,10 +73,38 @@ export default Ember.Mixin.create({
       assert('shouldCache needs to be a boolean value', typeof(shouldCache) === 'boolean');
 
       if (shouldCache) {
-        queriesForModel[cacheReadStringifiedQuery] = queryPromise;
+        this._scopedQueryCache.cache.add(modelName, cacheReadStringifiedQuery, queryPromise);
       }
 
       return results;
     });
+  },
+
+  scheduleCacheDecay(delay = CACHE_DECAY_TIMEOUT, callback) {
+    if (this._scopedQueryCache.decayId) {
+      this.cancelCacheDecay();
+    }
+
+    this._scopedQueryCache.decayId = run.later(() => {
+      if (this.isDestroying || this.isDestroyed) {
+        return;
+      }
+
+      this._scopedQueryCache = this._createScopedQueryCache();
+
+      callback && typeof callback === 'function' && callback();
+    }, delay);
+  },
+
+  cancelCacheDecay() {
+    run.cancel(this._scopedQueryCache.decayId);
+    this._scopedQueryCache.decayId = null;
+  },
+
+  _createScopedQueryCache() {
```

Comment:
this method appears to not rely on our mutate state (not does it seems configurable) we should likely remove it from the prototype and make it a helper function (in the same module).

This will prevent this mixing from needlessly increasing the interface it adds to things it is mixed into.

Also this would make unit testing this function quite simple

```js
function createCache() {
  return {
    cache: new Cache(),
    decayId: null
  }
}
```

But after reading this, My guess is that this QueryCache object actually just wants to be an ES6 class. with methods such ass `add` `remove` `schedule` `destroy`

```js
Mixin.create({

  init() {
    this.queryCache = new QueryCache(); // now `queryCache` is self contained, and its public API are also namespaced and available.
  }

  destroy() {
    this.queryCache.destroy();
    this._super(...arguments);
  }
})
```